Attempt at mapping google spans to opentelemetry ones better

euri10 opened the pull request:
Attempt at solving https://github.com/GoogleCloudPlatform/opentelemetry-operations-python/issues/89

it seems there are predefined labels as per https://cloud.google.com/trace/docs/reference/v1/rest/v1/projects.traces#get

I'm not sure it's working but maye a simple mapping like this would be ok ?

euri10 commented:
this PR seems relevant too, solved using a dictionary, 

https://github.com/open-telemetry/opentelemetry-dotnet/pull/453/files

T

aabmass commented:
@euri10 Thanks for the PR! I think an explicit mapping would be better in case of inconsistencies. Do you still have time to work on this PR? 😄 

Also an fyi, the `g.co/agent` attribute is for the telemetry library, not http user agent: https://github.com/GoogleCloudPlatform/opentelemetry-operations-python/blob/7b71183b8e6cffcdbc4805687b9546f409bf0885/opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_trace/__init__.py#L384-L391

euri10 commented on opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_trace/__init__.py:
```diff
@@ -361,6 +361,32 @@ def _extract_resources(resource: Resource) -> Dict[str, str]:
+    "http.path": "/http/path",
+    "http.redirected_url": "/http/redirected_url",
+    # https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/http.md#common-attributes
+    "http.request_content_length": "/http/request/size",
+    "http.response_content_length": "/http/response/size",
```
Please note that I used for those 2 a different mapping than the rest, could have been `http.request_size` but  I saw instead `http.request_content_length` which seems to fit nicely, not too sure about it though

aabmass replied:
Ya it's not totally clear. I think this is fine for now until I can get clarification from the Cloud Trace folks

euri10 commented:
> @euri10 Thanks for the PR! Do you still have time to work on this PR?  I think an explicit mapping would be better in case of inconsistencies.
> 
> Also an fyi, the `g.co/agent` attribute is for the telemetry library, not http user agent:
> 
> https://github.com/GoogleCloudPlatform/opentelemetry-operations-python/blob/7b71183b8e6cffcdbc4805687b9546f409bf0885/opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_trace/__init__.py#L384-L391

I redid it as a mapping, ready now, I left a comment on one of the mapping though

aabmass commented on opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_trace/__init__.py:
```diff
@@ -361,36 +361,30 @@ def _extract_resources(resource: Resource) -> Dict[str, str]:
+    "component": "/component",
+    "error.message": "/error/message",
+    "error.name": "/error/name",
+    "http.client_city": "/http/client_city",
+    "http.client_country": "/http/client_country",
+    "http.client_protocol": "/http/client_protocol",
+    "http.client_region": "/http/client_region",
+    "http.host": "/http/host",
+    "http.method": "/http/method",
+    "http.path": "/http/path",
+    "http.redirected_url": "/http/redirected_url",
+    # https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/http.md#common-attributes
+    "http.request_content_length": "/http/request/size",
+    "http.response_content_length": "/http/response/size",
+    "http.route": "/http/route",
+    "http.status_code": "/http/status_code",
+    "http.url": "/http/url",
+    "http.user_agent": "/http/user_agent",
```
A few tweaks and we should remove the ones not in the specification (https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/http.md)

```suggestion
    # this one might be http.flavor? I'm not sure
    "http.scheme": "/http/client_protocol",
    "http.host": "/http/host",
    "http.method": "/http/method",
    # https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/http.md#common-attributes
    "http.request_content_length": "/http/request/size",
    "http.response_content_length": "/http/response/size",
    "http.route": "/http/route",
    "http.status_code": "/http/status_code",
    "http.url": "/http/url",
    "http.user_agent": "/http/user_agent",
```

aabmass commented on opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_trace/__init__.py:
```diff
@@ -361,36 +361,30 @@ def _extract_resources(resource: Resource) -> Dict[str, str]:
+    "http.status_code": "/http/status_code",
+    "http.url": "/http/url",
+    "http.user_agent": "/http/user_agent",
+    "pid": "/pid",
```
This one may be available [in the Resource `process.pid`](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/resource/semantic_conventions/process.md#process).

Feel free to just leave it out for now
```suggestion
```

euri10 replied:
I added it

aabmass commented on opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_trace/__init__.py:
```diff
@@ -361,36 +361,30 @@ def _extract_resources(resource: Resource) -> Dict[str, str]:
+    "http.url": "/http/url",
+    "http.user_agent": "/http/user_agent",
+    "pid": "/pid",
+    "stacktrace": "/stacktrace",
```
Exceptions are added as [an `Event` called "exception" in OTel](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/exceptions.md) where you can get the `exception.stacktrace` attribute. Feel free to skip for now :)
```suggestion
```

euri10 replied:
see https://github.com/GoogleCloudPlatform/opentelemetry-operations-python/pull/90#discussion_r517961230

aabmass commented on opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_trace/__init__.py:
```diff
@@ -403,7 +397,8 @@ def _extract_attributes(
     for key, value in attrs.items() if attrs else []:
         key = _truncate_str(key, 128)[0]
-        key = _key_predefined(key)
+        if key in LABELS_MAPPING.keys():
```
```suggestion
        if key in LABELS_MAPPING:
```

aabmass commented on opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_trace/__init__.py:
```diff
@@ -364,6 +364,32 @@ def _extract_resources(resource: Resource) -> Dict[str, str]:
+    "http.user_agent": "/http/user_agent",
+    "pid": "/pid",
+    "stacktrace": "/stacktrace",
+    "tid": "/tid",
```
```suggestion
```

aabmass reviewed (commented):
Could you add a test case for this change as well?

euri10 commented on opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_trace/__init__.py:
```diff
@@ -361,36 +361,30 @@ def _extract_resources(resource: Resource) -> Dict[str, str]:
+LABELS_MAPPING = {
+    # "agent": "/agent",
+    "component": "/component",
+    "error.message": "/error/message",
+    "error.name": "/error/name",
```
one could argue the following could be used as well, not sure in fact as the google documentation is scarce about the meaning they give to the labels.

`exception.message` for `/error/message`
`exception.type` for `/error/name`
`exception.stacktrace` for `/stacktrace`

https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/exceptions.md#attributes

aabmass replied:
I think `span.status` would be most appropriate for mapping to `/error/message` and `/error/name` because it is a single field meant to classify the *entire* span, while you can have multiple exception `Event`s and they don't necessarily mean an error.

I'm not sure how that fits with `/stacktrace`, since you can have multiple exceptions. This could use some clarification from someone on Cloud Trace.

euri10 replied:
ok, I removed them

euri10 commented:
wip for a test

euri10 commented:
ok @aabmass I think I'm ready for a 2nd pass !

1. I added a test, I'm very new to the lib so hopefully I did it ok
2. removed all non-mapped entries
3. kept some entries that seems to fit ok with the spec, unfortunately I didnt find a google doc explaining clearly the intent of each entry so there's probably room for improvement.

aabmass commented on opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_trace/__init__.py:
```diff
@@ -364,6 +364,27 @@ def _extract_resources(resource: Resource) -> Dict[str, str]:
 
+LABELS_MAPPING = {
+    # https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/exceptions.md#attributes
+    "exception.message": "/error/message",
+    "exception.type": "/error/name",
+    "exception.stacktrace": "/stacktrace",
```
These are not set in span attributes directly, they will be in an `Event` named "exception".

aabmass commented on opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_trace/__init__.py:
```diff
@@ -364,6 +364,27 @@ def _extract_resources(resource: Resource) -> Dict[str, str]:
+    "http.url": "/http/url",
+    "http.user_agent": "/http/user_agent",
+    # https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/resource/semantic_conventions/process.md#process
+    "process.pid": "/pid",
```
Same with this one, it is not in attributes but in `span.resource`

euri10 replied:
removed as well

aabmass commented on opentelemetry-exporter-google-cloud/tests/test_cloud_trace_exporter.py:
```diff
@@ -84,6 +84,76 @@ def setUpClass(cls):
                 "int_key": AttributeValue(int_value=123),
             }
         )
+        cls.attributes_labels_mapping = {
+            "exception.message": "exception message",
+            "exception.type": "ValueError",
+            "exception.stacktrace": "exception stacktrace",
+            "http.scheme": "http",
+            "http.host": "172.19.0.4:8000",
+            "http.method": "POST",
+            "http.request_content_length": 321,
+            "http.response_content_length": 123,
+            "http.route": "/fuzzy/search",
+            "http.status_code": 200,
+            "http.url": "http://172.19.0.4:8000/fuzzy/search",
+            "http.user_agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.116 Safari/537.36",
+            "process.pid": 1111,
+        }
+        cls.extracted_attributes_labels_mapping = ProtoSpan.Attributes(
```
nit since these are only used in the `test_extract_label_mapping_attributes` case, I'd just instantiate them there for clarity

euri10 replied:
done

aabmass commented on opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_trace/__init__.py:
```diff
@@ -374,6 +389,8 @@ def _extract_attributes(
     for key, value in attrs.items() if attrs else []:
         key = _truncate_str(key, 128)[0]
+        if key in LABELS_MAPPING:
+            key = LABELS_MAPPING.get(key, key)
```
nit covered by if statement already
```suggestion
            key = LABELS_MAPPING[key]
```

euri10 replied:
yep corrected, but now CI failed, I added that ugly get in the 11st place because of it, not familiar with pylint but surely one of theses cases where we could ignore the linter

aabmass reviewed (approved):
Awesome one last nit, otherwise LGTM

aabmass reviewed (changes requested):
Oh, can you also update the changelog? :)

euri10 commented:
> Oh, can you also update the changelog? :)

done :)

aabmass reviewed (approved):
Thanks again for the PR @euri10 😄  will merge once the conflicts are resolved!

euri10 commented:
> Thanks again for the PR @euri10  will merge once the conflicts are resolved!

sorry I didnt see I was behind master !
